doc/design/008-state-proof-pluggable-parsing/README.md

```diff
@@ -0,0 +1,127 @@
+## Legend
+There are some types of requests to Nodes in the Pool allowing to use StateProof optimization in Client-Node communication.
+Instead of sending requests to all nodes in the Pool, a client can send a request to a single Node and expect StateProof signed by BLS multi-signature.
+
+BLS multi-signature (BLS MS) guaranties that there was a consensus of Nodes which signed some State identified by State RootHash.
+StateProof (SP) is small amount of data which allows to verify particular values against RootHash.
+Combination of BLS MS and SP allows clients to be sure about response of single node is a part of State signed by enough Nodes. 
+
+## Goals
+Libindy allows to create and send custom transaction via plugged interface.
+It is nice to have a way to support BLS MS and SP verification for these plugged transactions.
+
+Implementation of math for SP verification is a bit complicate for plugin logic.
+Therefore libindy should perform all math calculation inside.
+A plugin should provide handler to parse custom reply to fixed data structure.
+
+## API
+The signature of the handler is described below together with custom `free` call to deallocate result data.
+
+```rust
+extern fn custom_transaction_parser(reply_from_node: *const c_char, parsed_sp: *mut *const c_char) -> ErrCode;
+extern fn custom_free(data: *mut c_char) -> ErrCode;
```

Comment:
Which data is supposed to free? In `parsed_sp`?

Reply:
```custom_transaction_parser```

This handler returns a null terminated string with result of request parsing by writing string pointer to memory located by ``````parsed_sp```` pointer.

Memory to store null terminated string will be allocated by plugin, so plugin should provide also handler for de-allocation of this memory.

Reply:
yes, `parsed_sp` is out parameter for parser and memory will be allocated by plugin for this pointer. `custom_free` will de-allocate this buffer